google/cloud/internal/url_encode.cc

```diff
@@ -24,35 +25,68 @@ namespace internal {
 
 namespace {
 
-using UrlMappings =
-    std::array<std::pair<absl::string_view, absl::string_view>, 25>;
-
-constexpr auto kEncodeMappings = UrlMappings{{
-    {" ", "%20"}, {"\"", "%22"}, {"#", "%23"},  {"$", "%24"}, {"%", "%25"},
-    {"&", "%26"}, {"+", "%2B"},  {",", "%2C"},  {"/", "%2F"}, {":", "%3A"},
-    {";", "%3B"}, {"<", "%3C"},  {"=", "%3D"},  {">", "%3E"}, {"?", "%3F"},
-    {"@", "%40"}, {"[", "%5B"},  {"\\", "%5C"}, {"]", "%5D"}, {"^", "%5E"},
-    {"`", "%60"}, {"{", "%7B"},  {"|", "%7C"},  {"}", "%7D"}, {"\177", "%7F"},
-}};
-
-// Until we require C++17, it isn't worth trying to eliminate the duplication
-// using a constexpr function to produce kDecodeMappings from kEncodeMappings.
-constexpr auto kDecodeMappings = UrlMappings{{
-    {"%20", " "}, {"%22", "\""}, {"%23", "#"},  {"%24", "$"}, {"%25", "%"},
-    {"%26", "&"}, {"%2B", "+"},  {"%2C", ","},  {"%2F", "/"}, {"%3A", ":"},
-    {"%3B", ";"}, {"%3C", "<"},  {"%3D", "="},  {"%3E", ">"}, {"%3F", "?"},
-    {"%40", "@"}, {"%5B", "["},  {"%5C", "\\"}, {"%5D", "]"}, {"%5E", "^"},
-    {"%60", "`"}, {"%7B", "{"},  {"%7C", "|"},  {"%7D", "}"}, {"%7F", "\177"},
-}};
+std::string ToHex(int i) {
+  std::stringstream stream;
+  stream << "%" << std::setfill('0') << std::setw(2) << std::uppercase
+         << std::hex << i;
+  return stream.str();
+}
+
+std::string Escape(char c) {
+  static auto const* const kDict = []() {
+    static std::unordered_map<char, std::string> m;
+    for (auto i = 0; i < 32; ++i) {
+      m[static_cast<char>(i)] = ToHex(i);
+    }
+    for (auto c : {
+             ' ', '\"', '#', '$', '%', '&',  '+', ',', '/', ':', ';', '<',
+             '=', '>',  '?', '@', '[', '\\', ']', '^', '`', '{', '|', '}',
+         }) {
+      m[c] = ToHex(static_cast<int>(c));
+    }
+    for (auto i = 127; i < 256; ++i) {
+      m[static_cast<char>(i)] = ToHex(i);
+    }
```

Comment:
Consider using `!std::isprint()` to determine the other chars that need to be escaped.